src/Libraries/DSCPython/DSCPythonCodeCompletionProviderCore.cs

```diff
@@ -0,0 +1,1292 @@
+            {
+                throw new ArgumentException("parameter n must be of type NameSpaceTracker");
+            }
+            var nst = nameSpaceTracker as NamespaceTracker;
+            object varOutput;
+
+            var periodIndex = name.IndexOf('.');
+            if (periodIndex == -1)
+            {
+                if (nst.TryGetValue(name, out varOutput))
+                {
+                    return varOutput;
+                }
+                return null;
+            }
+
+            var currentName = name.Substring(0, periodIndex);
+            var theRest = name.Substring(periodIndex + 1);
+
+            if (nst.TryGetValue(currentName, out varOutput))
+            {
+                if (varOutput is NamespaceTracker)
+                {
+                    return LookupMember(theRest, varOutput as NamespaceTracker);
+                }
+            }
+            return null;
+        }
+
+        /// <summary>
+        /// Recursively lookup a variable in the _scope
+        /// </summary>
+        /// <param name="name">A name for a type, possibly delimited by periods.</param>
+        /// <returns>The type as an object</returns>
+        public object LookupMember(string name)
+        {
+            object varOutput = null;
+            object currentOutput = null;
+            var periodIndex = name.IndexOf('.');
+
+            IntPtr gs = PythonEngine.AcquireLock();
+            try
+            {
+                using (Py.GIL())
+                {
+                    if (periodIndex == -1)
+                    {
+                        if (Scope.TryGet(name, out varOutput))
+                        {
+                            return varOutput;
+                        }
+                        return null;
+                    }
+
+                    var currentName = name.Substring(0, periodIndex);
+                    var theRest = name.Substring(periodIndex + 1);
+                    Scope.TryGet(currentName, out currentOutput);
+
+                    if (currentOutput != null)
+                    {
+                        if (currentOutput is NamespaceTracker)
+                        {
+                            return LookupMember(theRest, varOutput as NamespaceTracker);
+                        }
+                    }
+                    return null;
+                }
+            }
+            finally
+            {
+                PythonEngine.ReleaseLock(gs);
+            }
+        }
+
+        private PythonEngine engine;
+        /// <summary>
+        /// The engine used for autocompletion.  This essentially keeps
+        /// track of the state of the editor, allowing access to variable types and
+        /// imported symbols.
+        /// </summary>
+        public object Engine
+        {
+            get { return engine; }
+            set { engine = (PythonEngine)value; }
+        }
+        private PyScope scope;
+
+        internal string globalScopeName = "globalScope";
```

Comment:
this is really strange - you're not using this as a globalScopeName, you're using it as a unique scope per completion provider. I think this could be named more clearly.

I have some thoughts on how to get rid of this completely, will try to expand below.

Reply:
Changed the variable name accordingly.